graylog2-web-interface/src/views/components/widgets/ExtraDropdownWidgetActions.tsx

```diff
@@ -26,42 +26,77 @@ import useSendTelemetry from 'logic/telemetry/useSendTelemetry';
 import { TELEMETRY_EVENT_TYPE } from 'logic/telemetry/Constants';
 import { getPathnameWithoutId } from 'util/URLUtils';
 import useLocation from 'routing/useLocation';
+import type { ActionComponents } from 'views/components/actions/ActionHandler';
+import type { WidgetActionType, WidgetAction } from 'views/components/widgets/Types';
+import generateId from 'logic/generateId';
 
+type SetComponents = (value: ((prevState: ActionComponents) => ActionComponents) | ActionComponents) => void;
 type Props = {
   widget: Widget;
+  setComponents: SetComponents;
 };
 
-const ExtraDropdownWidgetActions = ({ widget }: Props) => {
+const createHandler = (action: WidgetActionType, setComponents: SetComponents): WidgetAction => {
+  if (action.action) {
+    return action.action;
+  }
+  if (action.component) {
+    const ActionComponent = action.component;
+
+    return (widget, contexts) => () => {
+      const id = generateId();
+
+      const onClose = () => setComponents(({ [id]: _, ...rest }) => rest);
+      const renderedComponent = (
+        <ActionComponent key={action.type} widget={widget} contexts={contexts} onClose={onClose} />
+      );
+
+      setComponents((actionComponents) => ({
+        [id]: renderedComponent,
+        ...actionComponents,
+      }));
+
+      return Promise.resolve();
+    };
+  }
+
+  return () => () => Promise.resolve();
+};
```

Comment:
The use case of rendering e.g. modals from inside a menu item, seems to be a repeating pattern. Currently we solve this differently in other cases, for example in `DashboardActions`.

Maybe we can abstract the related logic and reuse it?

Reply:
Yeah, I thought about it too and looked into the other places, but it requires a bit more effort for a good abstraction. It is on my list though, for a future PR. Thanks!